### ISSUE TITLE
[bug] build order produces not buildable build arguments in specific override scenario

### ISSUE THREAD

**Issue by nicosmd** [conan-io/conan#16116]
### Describe the bug

Hi!

I'm using conan 2.2.2 and I'm currently having an issue with the output of `conan graph build-order` in a specific override scenario. The issue appears as soon as I try to build packages using the build order output. I have created a repository to demonstrate and reproduce the issue: https://github.com/nicosmd/conan2-build-order-override-failed

Please let me know if you need further information. 

Thank you very much!

### How to reproduce it

clone https://github.com/nicosmd/conan2-build-order-override-failed and run `build.sh`

**Comment by memsharded** [comment]
Hi @nicosmd 

Thanks very much for reporting this.

I haven't dropped this, this is indeed a bug, but it is proving a very challenging it. I have been working on this since then, and I'll try to fix it for next 2.3

**Comment by nicosmd** [comment]
Hey @memsharded,

thank you for the update and for working on this. If there is anything I can do to support you on this, please let me know!

**Comment by memsharded** [comment]
Update: this is the ticket that I have been investing more time in a long time 😅 

It happened to be quite complicated, I had to try many different approaches, failed in all of them. Finally a relatively easy solution seems workable in https://github.com/conan-io/conan/pull/16235, I am proposing it for next patch 2.3.1 release. I'll probably add a few more tests, but in case you want to give it a try and run from source for your project and see how it goes.

**Comment by memsharded** [comment]
Hi @nicosmd 

Did you have a chance to test the solution in https://github.com/conan-io/conan/pull/16235? I think it is almost ready to be able to release it in 2.3.1, but more feedback would be very useful.

**Comment by nicosmd** [comment]
Hey @memsharded ,

I have just tested the solution with the repository I have created and it works without any issues!

**Comment by memsharded** [comment]
Excellent, thanks for the feedback!

**Comment by nicosmd** [comment]
Maybe I have found a regression but I'm not sure yet. In production I am using [replace_requires] to override some dependencies to be found in a sysroot. Using the bugfix branch I'm getting the error `ERROR: Requirement 'openssl/system' not in lockfile` while creating the build order json using the created lockfiles. I think that openssl itself is overriden from version `openssl/1.1.1s` to `openssl/1.1.1t` in one downstream dependency so maybe it is the combination between usual override and replace_requires. I'm not getting this error using conan 2.3.0. The lockfiles using 2.3.0 and the fix branch both contain openssl/system as a requirement and the rest is almost identical. The only difference is:

2.3.0:
```
"overrides": {
...
"openssl/1.1.1s": [
            "openssl/1.1.1t",
            null
        ]
...
```
fix branch:
```bash
"overrides": {
...
"openssl/1.1.1s": [
            "openssl/1.1.1t#1fd1eb221ddba2f48d5abf4d9a50058e",
            null
        ]
...
```

 I would need some time to reproduce and create a minimal demo for the issue if needed.

Do you have any idea what could went wrong?

**Comment by memsharded** [comment]
No idea, I am trying to reproduce, adding further tests like https://github.com/conan-io/conan/pull/16235/commits/70e58cfe16d252b6b7f51c04002a8f6f1703c0b9, but so far I haven't manage to make it fail.

**Comment by memsharded** [comment]
I have managed to find a small issue that seems related to your case, and did a fix in https://github.com/conan-io/conan/pull/16235/commits/bdb8bd06f98849a93790159e7481b94b9dbbff83, including a very large test.

If you can update the code and try again, please let me know.

**Comment by nicosmd** [comment]
Hey @memsharded thank you very much! The issue disappeared now :)

**Comment by czoido** [comment]
Closed by https://github.com/conan-io/conan/pull/16235 to be released in 2.3.1 very soon
